lib/Dialects/LinalgTransform/Transforms/TrackingCSE.cpp

```diff
@@ -0,0 +1,284 @@
+      return false;
+    return OperationEquivalence::isEquivalentTo(
+        const_cast<Operation *>(lhsC), const_cast<Operation *>(rhsC),
+        /*mapOperands=*/OperationEquivalence::exactValueMatch,
+        /*mapResults=*/OperationEquivalence::ignoreValueEquivalence,
+        OperationEquivalence::IgnoreLocations);
+  }
+};
+
+// -------------------------------------------------------------------------- //
+// End of code below is taken from CSE.
+// -------------------------------------------------------------------------- //
+
+struct TrackingCSE {
+  TrackingCSE(DominanceInfo *domInfo,
+              const DenseMap<Value, Operation *> &trackedOperations)
+      : domInfo(domInfo) {
+    for (auto &pair : trackedOperations)
+      trackedOps.insert(pair.second);
+  }
+
+  void simplify(Operation *root);
+
+  // ------------------------------------------------------------------------ //
+  // The code below is adapted from CSE.
+  // ------------------------------------------------------------------------ //
+
+  /// Shared implementation of operation elimination and scoped map definitions.
+  using AllocatorTy = llvm::RecyclingAllocator<
+      llvm::BumpPtrAllocator,
+      llvm::ScopedHashTableVal<Operation *, Operation *>>;
+  using ScopedMapTy = llvm::ScopedHashTable<Operation *, Operation *,
+                                            SimpleOperationInfo, AllocatorTy>;
+
+  /// Represents a single entry in the depth first traversal of a CFG.
+  struct CFGStackNode {
+    CFGStackNode(ScopedMapTy &knownValues, DominanceInfoNode *node)
+        : scope(knownValues), node(node), childIterator(node->begin()),
+          processed(false) {}
+
+    /// Scope for the known values.
+    ScopedMapTy::ScopeTy scope;
+
+    DominanceInfoNode *node;
+    DominanceInfoNode::const_iterator childIterator;
+
+    /// If this node has been fully processed yet or not.
+    bool processed;
+  };
+
+  /// Attempt to eliminate a redundant operation. Returns success if the
+  /// operation was marked for removal, failure otherwise.
+  LogicalResult simplifyOperation(ScopedMapTy &knownValues, Operation *op,
+                                  bool hasSSADominance);
+  void simplifyBlock(ScopedMapTy &knownValues, Block *bb, bool hasSSADominance);
+  void simplifyRegion(ScopedMapTy &knownValues, Region &region);
+
+private:
+  /// Operations marked as dead and to be erased.
+  std::vector<Operation *> opsToErase;
+  DominanceInfo *domInfo = nullptr;
+  llvm::SmallPtrSet<Operation *, 8> trackedOps;
+};
+} // end anonymous namespace
+
+/// Attempt to eliminate a redundant operation.
+LogicalResult TrackingCSE::simplifyOperation(ScopedMapTy &knownValues,
+                                             Operation *op,
+                                             bool hasSSADominance) {
+  // Don't simplify terminator operations.
+  if (op->hasTrait<OpTrait::IsTerminator>())
+    return failure();
+
+  // Never remove tracked ops.
+  if (trackedOps.count(op)) {
```

Comment:
could you please make it very visible with (comments) that this is the only place in the TrackingCSE that differs from the original impl?

Better yet, would it be possible to paste the original CSE and just change it slightly so that it can be derievd and the extra bahavior just be injected with a few lines?